src/main/java/org/opensearch/neuralsearch/processor/InferenceProcessor.java

```diff
@@ -119,6 +125,83 @@ public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Ex
         }
     }
 
+    abstract void doBatchExecute(List<String> inferenceList, Consumer<List<?>> handler, Consumer<Exception> onException);
+
+    @Override
+    public void batchExecute(List<IngestDocumentWrapper> ingestDocumentWrappers, Consumer<List<IngestDocumentWrapper>> handler) {
+        if (CollectionUtils.isEmpty(ingestDocumentWrappers)) {
+            handler.accept(Collections.emptyList());
+            return;
+        }
+
+        List<DataForInference> dataForInferences = getDataForInference(ingestDocumentWrappers);
+        List<String> inferenceList = constructInferenceTexts(dataForInferences);
+        if (inferenceList.isEmpty()) {
+            handler.accept(ingestDocumentWrappers);
+            return;
+        }
+        doBatchExecute(inferenceList, results -> {
+            int startIndex = 0;
+            for (DataForInference dataForInference : dataForInferences) {
+                if (dataForInference.getIngestDocumentWrapper().getException() != null || dataForInference.getInferenceList().isEmpty()) {
+                    continue;
+                }
+                List<?> inferenceResults = results.subList(startIndex, startIndex + dataForInference.getInferenceList().size());
+                startIndex += dataForInference.getInferenceList().size();
+                setVectorFieldsToDocument(
+                    dataForInference.getIngestDocumentWrapper().getIngestDocument(),
+                    dataForInference.getProcessMap(),
+                    inferenceResults
+                );
+            }
+            handler.accept(ingestDocumentWrappers);
+        }, exception -> {
```

Comment:
Is it possible for parts of documents failed but others success? For example, for local deployment case, if we do the inference one by one (I didn't confirm it but it should be), is it possible?

Reply:
Here is the code for local model, https://github.com/opensearch-project/ml-commons/blob/caf1d65fb3372c4a39c8d21545016ffdbf6c1c9a/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/TextEmbeddingModel.java#L43

we predict them one by one and if any of the prediction throws exception, the whole batch predict throw exception, so it's either success or all fail.

For remote inference, partial failure will cause inference exception callback triggered https://github.com/opensearch-project/ml-commons/blob/94a113da8a2d1e28a84ba7a422b43287ac00448e/ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java#L183.

Since the logic of how processors invoke ml-commons remains the same, the error handling logic also remains the same.